resources/type-system/inference.md

```diff
@@ -1210,6 +1247,674 @@ of steps:
+      `e_i`, in context `K_i`.
+
+  - For each `e_i` in stage _k_ that _does_ take the form of a
+    _&lt;functionExpression&gt;_ enclosed in zero or more parentheses, in order
+    of increasing _i_:
+
+    - Let `m_i_preliminary` be the result of performing expression inference on
+      `e_i`, in context `K_i`.
+
+  - _Note that a property of argument partitioning is that arguments that are
+    not function literal expressions are always placed in stage zero, so this
+    has the effect that all arguments that are not function expressions are type
+    inferred first, in the order in which they appear in source code, followed
+    by all arguments that __are__ function expressions, possibly in a staged
+    fashion._
+
+  - If `F` has at least one formal type parameter, and there are zero type
+    arguments `T` (_in other words, generic type inference is being performed_),
+    then:
+
+    - For each `e_i` in stage _k_:
+
+      - Let `S_i` be the static type of `m_i_preliminary`.
+
+      - Using subtype constraint generation, attempt to match `S_i` as a subtype
+        of `K_i` with respect to the list of type variables `{X_1, X_2, ...}`.
+
+        - If this succeeds, then accumulate the resulting list of type
+          constraints into `C`.
+
+        - Otherwise, leave `C` unchanged.
+
+    - Update `{U_1, U_2, ...}` as follows:
+
+      - If _k_ is not the last stage in the argument partitioning, then let
+        `{V_1, V_2, ...}` be the constraint solution for the set of type
+        variables `{X_1, X_2, ...}` with respect to the constaint set `C`, with
+        partial solution `{U_1, U_2, ...}`. Replace `{U_1, U_2, ...}` with
+        `{V_1, V_2, ...}`. _This step is known as "horizontal inference", since
+        it has the effect of passing type information __across__ the syntax tree
+        from the static type of some arguments to the context of other
+        arguments._
+
+      - Otherwise (_k_ __is__ the last stage in the argument partitioning), let
+        `{V_1, V_2, ...}` be the _grounded_ constraint solution for the set of
+        type variables `{X_1, X_2, ...}` with respect to the constaint set `C`,
+        with partial solution `{U_1, U_2, ...}`. Replace `{U_1, U_2, ...}` with
+        `{V_1, V_2, ...}`. _This step is known as "upwards inference", since it
+        has the effect of passing type information __up__ the syntax tree from
+        the static type of arguments to the inferred type arguments of the
+        invocation._
+
+- _Note that at this point, all entries in `{U_1, U_2, ...}` have been obtained
+  either from a _grounded_ constraint solution or from explicit types in the
+  source code, so they are guaranteed to be proper types, not type schemas._
+
+- For each `e_i`:
+
+  - Let `K_i` be the result of substituting `{U_1, U_2, ...}` for `{X_1, X_2,
+    ...}` in `P_i`. _Note that this is now guaranteed to be a proper type, not a
```

Comment:
At some point up above (in the case that the function type was missing) you set `P_i` to `_`, in which case this is not true.  It's not clear to me whether you ever end up at this particular point in the algorithm if you are in the "missing function type" case, but this is an example of why I think it might be much clearer to just split out the three cases (dynamic, explicit type arguments, inferred type arguments) into three sections instead of trying to interleave them.

Reply:
Hmm, you are right; this doesn’t make sense. And I agree that this is a good argument for splitting the algorithm into three cases.